[
   {
      "body" : "What operating system are you using?\r\nHow much memory/disk space is available?",
      "created_at" : "2016-10-24T01:05:43Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255627918",
      "id" : 255627918,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T01:05:43Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255627918",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/863730?v=3",
         "events_url" : "https://api.github.com/users/fanquake/events{/privacy}",
         "followers_url" : "https://api.github.com/users/fanquake/followers",
         "following_url" : "https://api.github.com/users/fanquake/following{/other_user}",
         "gists_url" : "https://api.github.com/users/fanquake/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/fanquake",
         "id" : 863730,
         "login" : "fanquake",
         "organizations_url" : "https://api.github.com/users/fanquake/orgs",
         "received_events_url" : "https://api.github.com/users/fanquake/received_events",
         "repos_url" : "https://api.github.com/users/fanquake/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/fanquake/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/fanquake/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/fanquake"
      }
   },
   {
      "body" : "> Such as running out of memory on a VPS\r\n\r\nIf that's the case your filesystem may be corrupt, and of course data on it as well. \r\nYou should check it with fsck and make sure it's sound before bitcoind is suspected.",
      "created_at" : "2016-10-24T05:30:08Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255650955",
      "id" : 255650955,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T05:30:08Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255650955",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/13134193?v=3",
         "events_url" : "https://api.github.com/users/unsystemizer/events{/privacy}",
         "followers_url" : "https://api.github.com/users/unsystemizer/followers",
         "following_url" : "https://api.github.com/users/unsystemizer/following{/other_user}",
         "gists_url" : "https://api.github.com/users/unsystemizer/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/unsystemizer",
         "id" : 13134193,
         "login" : "unsystemizer",
         "organizations_url" : "https://api.github.com/users/unsystemizer/orgs",
         "received_events_url" : "https://api.github.com/users/unsystemizer/received_events",
         "repos_url" : "https://api.github.com/users/unsystemizer/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/unsystemizer/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/unsystemizer/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/unsystemizer"
      }
   },
   {
      "body" : "A sudden shutdown (crash/kill) of bitcoind may lead to database corruption (which results in a re-sync from the scratch). I observed these types of corruptions often on VPS.\r\nFeatures like https://github.com/bitcoin/bitcoin/issues/8037 could be a relief in such cases... \r\n\r\nIMO applications with heavy database-interaction like bitcoind (UTXO set interaction) tend to loose integrity in force shutdown (crash/kill) situations.",
      "created_at" : "2016-10-24T06:11:37Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255655783",
      "id" : 255655783,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T06:11:37Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255655783",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/178464?v=3",
         "events_url" : "https://api.github.com/users/jonasschnelli/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jonasschnelli/followers",
         "following_url" : "https://api.github.com/users/jonasschnelli/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jonasschnelli/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jonasschnelli",
         "id" : 178464,
         "login" : "jonasschnelli",
         "organizations_url" : "https://api.github.com/users/jonasschnelli/orgs",
         "received_events_url" : "https://api.github.com/users/jonasschnelli/received_events",
         "repos_url" : "https://api.github.com/users/jonasschnelli/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jonasschnelli/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jonasschnelli/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jonasschnelli"
      }
   },
   {
      "body" : "@fanquake, this is Fedora 24 with 1GB of RAM. That said, this is 100% reproducible for me with the given settings. It shouldn't depend on available RAM.\r\n\r\n@unsystemizer, the error is repeatable, and I don't think there is any situation where an OOM would lead to FS corruption.\r\n\r\nNote here that the issue isn't the database corruption, or the requirement that the re-sync happen - that's fine. The issue is that bitcoind hits an assert and exits immediately, rather than automatically starting a re-sync from scratch.",
      "created_at" : "2016-10-24T06:41:45Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255659892",
      "id" : 255659892,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T06:41:45Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255659892",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/1308314?v=3",
         "events_url" : "https://api.github.com/users/tdaede/events{/privacy}",
         "followers_url" : "https://api.github.com/users/tdaede/followers",
         "following_url" : "https://api.github.com/users/tdaede/following{/other_user}",
         "gists_url" : "https://api.github.com/users/tdaede/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/tdaede",
         "id" : 1308314,
         "login" : "tdaede",
         "organizations_url" : "https://api.github.com/users/tdaede/orgs",
         "received_events_url" : "https://api.github.com/users/tdaede/received_events",
         "repos_url" : "https://api.github.com/users/tdaede/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/tdaede/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/tdaede/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/tdaede"
      }
   },
   {
      "body" : "@tdaede: the assertion you hit is very likely caused by a database corruption (block index).",
      "created_at" : "2016-10-24T06:55:55Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255661961",
      "id" : 255661961,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T06:55:55Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255661961",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/178464?v=3",
         "events_url" : "https://api.github.com/users/jonasschnelli/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jonasschnelli/followers",
         "following_url" : "https://api.github.com/users/jonasschnelli/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jonasschnelli/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jonasschnelli",
         "id" : 178464,
         "login" : "jonasschnelli",
         "organizations_url" : "https://api.github.com/users/jonasschnelli/orgs",
         "received_events_url" : "https://api.github.com/users/jonasschnelli/received_events",
         "repos_url" : "https://api.github.com/users/jonasschnelli/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jonasschnelli/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jonasschnelli/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jonasschnelli"
      }
   },
   {
      "body" : "@tdaede - when you say it's repeatable, if you just restart bitcoind without fixing anything, I believe it's repeatable. It could also be repeatable if you shutdown the system while bitcoind is running and corrupt data the same or similar way. It doesn't mean it's a problem with Bitcoin Core. As I said you need to prove with fsck that the filesystem is sound. Even then it may be a problem with LevelDB or something that would have to be dealt with upstream.\r\n\r\nOOM can't lead to FS corruption: I disagree.\r\nhttps://unix.stackexchange.com/questions/12699/do-journaling-filesystems-guarantee-against-corruption-after-a-power-failure",
      "created_at" : "2016-10-24T07:02:39Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-255662906",
      "id" : 255662906,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-24T07:05:21Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/255662906",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/13134193?v=3",
         "events_url" : "https://api.github.com/users/unsystemizer/events{/privacy}",
         "followers_url" : "https://api.github.com/users/unsystemizer/followers",
         "following_url" : "https://api.github.com/users/unsystemizer/following{/other_user}",
         "gists_url" : "https://api.github.com/users/unsystemizer/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/unsystemizer",
         "id" : 13134193,
         "login" : "unsystemizer",
         "organizations_url" : "https://api.github.com/users/unsystemizer/orgs",
         "received_events_url" : "https://api.github.com/users/unsystemizer/received_events",
         "repos_url" : "https://api.github.com/users/unsystemizer/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/unsystemizer/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/unsystemizer/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/unsystemizer"
      }
   },
   {
      "body" : "@unsystemizer OOM isn't a power failure. It cannot cause filesystem corruption unless there are very serious kernel bugs.",
      "created_at" : "2016-10-26T03:17:59Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-256239361",
      "id" : 256239361,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-26T03:17:59Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/256239361",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/1095675?v=3",
         "events_url" : "https://api.github.com/users/luke-jr/events{/privacy}",
         "followers_url" : "https://api.github.com/users/luke-jr/followers",
         "following_url" : "https://api.github.com/users/luke-jr/following{/other_user}",
         "gists_url" : "https://api.github.com/users/luke-jr/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/luke-jr",
         "id" : 1095675,
         "login" : "luke-jr",
         "organizations_url" : "https://api.github.com/users/luke-jr/orgs",
         "received_events_url" : "https://api.github.com/users/luke-jr/received_events",
         "repos_url" : "https://api.github.com/users/luke-jr/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/luke-jr/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/luke-jr/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/luke-jr"
      }
   },
   {
      "body" : "Or in the hypervisor or h/w drivers sitting below the VM or somewhere else. It should still be shown that the filesystem is not corrupt, I think.",
      "created_at" : "2016-10-26T03:34:20Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-256241162",
      "id" : 256241162,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2016-10-26T03:34:20Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/256241162",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/13134193?v=3",
         "events_url" : "https://api.github.com/users/unsystemizer/events{/privacy}",
         "followers_url" : "https://api.github.com/users/unsystemizer/followers",
         "following_url" : "https://api.github.com/users/unsystemizer/following{/other_user}",
         "gists_url" : "https://api.github.com/users/unsystemizer/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/unsystemizer",
         "id" : 13134193,
         "login" : "unsystemizer",
         "organizations_url" : "https://api.github.com/users/unsystemizer/orgs",
         "received_events_url" : "https://api.github.com/users/unsystemizer/received_events",
         "repos_url" : "https://api.github.com/users/unsystemizer/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/unsystemizer/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/unsystemizer/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/unsystemizer"
      }
   },
   {
      "body" : "I think this is almost certainly not filesystem corruption. I can reproduce this failure mode by manually removing a single block from my block index when flushing to disk, and then starting bitcoind again. I hit the assert with this backtrace:\r\n\r\n```\r\n#0  CBlockIndex::GetAncestor (this=0x555556396400, height=8) at chain.cpp:105\r\n#1  0x000055555592b7e6 in CBlockIndex::BuildSkip (this=0x555556395ec0) at chain.cpp:123\r\n#2  0x000055555587de97 in LoadBlockIndexDB (chainparams=...) at validation.cpp:3526\r\n#3  0x00005555558800f9 in LoadBlockIndex (chainparams=...) at validation.cpp:3815\r\n#4  0x00005555555b1fb2 in AppInitMain (threadGroup=..., scheduler=...) at init.cpp:1428\r\n#5  0x0000555555583f8f in AppInit (argc=2, argv=0x7fffffffe568) at bitcoind.cpp:167\r\n#6  0x0000555555584668 in main (argc=2, argv=0x7fffffffe568) at bitcoind.cpp:196\r\n```\r\n\r\nIf I do anything that corrupts the block index in a more intrusive way (eg removing the pprev pointer or changing any of the other header fields), then we fail at a different point. The blockhash is no longer valid so we fail `CheckProofOfWork()`. Critically, `CBlockIndex::BuildSkip` in `LoadBlockIndexDB()` is called after `LoadBlockIndexGuts()`, which loads *all* of the block indexes from disk. So it looks to me like this failure mode can probably only be hit if all the indexes on disk are valid, but one or more block are missing.\r\n\r\n@tdaede are you able to upload debug.log? You suspect that this may be something to do with pruning. I've had quite a close read of that code and I can't see where it would cause us to lose block indexes or not flush them to disk. If you have a debug.log, it might help pin down where we're losing the block index.\r\n\r\nI'm also planning to open a PR which gives us slightly better diagnostics here by printing out the blockhash and height of the orphan block.",
      "created_at" : "2017-01-24T23:03:26Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-274967877",
      "id" : 274967877,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2017-01-24T23:03:26Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/274967877",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/1063656?v=3",
         "events_url" : "https://api.github.com/users/jnewbery/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jnewbery/followers",
         "following_url" : "https://api.github.com/users/jnewbery/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jnewbery/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jnewbery",
         "id" : 1063656,
         "login" : "jnewbery",
         "organizations_url" : "https://api.github.com/users/jnewbery/orgs",
         "received_events_url" : "https://api.github.com/users/jnewbery/received_events",
         "repos_url" : "https://api.github.com/users/jnewbery/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jnewbery/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jnewbery/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jnewbery"
      }
   },
   {
      "body" : "@jnewbery I had an offline conversation with @gmaxwell and apparently flushing is disabled during initial sync for speed, which means that the indexes can end up ahead of the block database. This was apparently done for speed - a simple fix would be to disable this optimization if it no longer gives significant speed gains.\r\n\r\nThere is also no way to fetch the missing block when pruning, so you have to start over.",
      "created_at" : "2017-01-25T03:37:38Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/issues/9001#issuecomment-275010615",
      "id" : 275010615,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/9001",
      "updated_at" : "2017-01-25T03:37:38Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/275010615",
      "user" : {
         "avatar_url" : "https://avatars.githubusercontent.com/u/1308314?v=3",
         "events_url" : "https://api.github.com/users/tdaede/events{/privacy}",
         "followers_url" : "https://api.github.com/users/tdaede/followers",
         "following_url" : "https://api.github.com/users/tdaede/following{/other_user}",
         "gists_url" : "https://api.github.com/users/tdaede/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/tdaede",
         "id" : 1308314,
         "login" : "tdaede",
         "organizations_url" : "https://api.github.com/users/tdaede/orgs",
         "received_events_url" : "https://api.github.com/users/tdaede/received_events",
         "repos_url" : "https://api.github.com/users/tdaede/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/tdaede/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/tdaede/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/tdaede"
      }
   }
]
